src/lightning/fabric/strategies/fsdp.py

```diff
@@ -594,29 +595,60 @@ def _set_world_ranks(self) -> None:
         rank_zero_only.rank = self.global_rank
 
 
-def _setup_activation_checkpointing(module: "FullyShardedDataParallel", layers: List[Type[Module]]) -> None:
+def _activation_checkpointing_kwargs(
+    activation_checkpointing: Optional[Union[Type[Module], List[Type[Module]]]] = None,
+    activation_checkpointing_policy: Optional["_POLICY"] = None,
+) -> Dict:
+    if activation_checkpointing is None and activation_checkpointing_policy is None:
+        return {}
+    if activation_checkpointing is not None and activation_checkpointing_policy is not None:
+        raise ValueError(
+            "You cannot set both `activation_checkpointing` and `activation_checkpointing_policy`. Use the latter."
+        )
+    if activation_checkpointing is not None:
+        if not _TORCH_GREATER_EQUAL_1_13:
+            raise ValueError("`activation_checkpointing` requires torch >= 1.13.0. HINT: `pip install -U torch`")
+        if isinstance(activation_checkpointing, list):
+            if len(activation_checkpointing) < 1:
+                raise ValueError("`activation_checkpointing` should not be an empty list")
+            classes = tuple(activation_checkpointing)
+        else:
+            classes = (activation_checkpointing,)
+        if _TORCH_GREATER_EQUAL_2_1:
+            rank_zero_deprecation(
+                f"`FSDPStrategy(activation_checkpointing={activation_checkpointing})` is deprecated, use "
+                "`FSDPStrategy(activation_checkpointing_policy=torch.distributed.fsdp.wrap.ModuleWrapPolicy"
+                f"({set(classes)}))` instead."
```

Comment:
Even after discussing offline, I still can't wrap my head around why we had to go back to this less convenient way. It essentially reverts the motivation behind https://github.com/Lightning-AI/lightning/issues/15825. This deprecation message shows that it is a step back:

1. The new argument name is longer
2. You have to extend it to two lines of code (import the ModuleWrapPolicy and update the argument)

Why don't we just support this:

```
auto_wrap_policy={classes}
activation_checkpointing_policy={classes}
```
internally we translate them to ModuleWrapPolicy({classes}) so you don't have to import it. What do you think?

Reply:
1. It's longer but `activation_checkpointing` is taken and `activation_checkpointing_policy` is consistent with `auto_wrap_policy` which takes exactly the same kind of input.
2. The user is most likely already defining a policy for `auto_wrap_policy`. For those who don't, they probably wouldn't mind the import because then they would need to manually wrap, which already requires that you import the wrapping functionality from PyTorch FSDP

> Why don't we just support this:

I would find that out of scope of this PR. I just want to implement what'ts minimally required for the new policy API. But feel free to propose it in a separate issue or PR. That sort of API would be similar to the proposal in https://github.com/Lightning-AI/lightning/issues/16991

Reply:
Note that https://github.com/Lightning-AI/lightning/issues/15825 was well motivated before PyTorch introduced policy support for activation checkpointing. Now it's only partially motivated: Lightning still takes care of the application itself, but the policy is left to the user

Reply:
I would feel more comfortable moving forward with this if we could discuss some ideas in here: https://github.com/Lightning-AI/lightning/issues/18067

Reply:
Commented